### PR TITLE
Fix Cross-Cache Deadlock with Notification Queue

### PR DESCRIPTION
**Pull request by dwcullop**
﻿## Fix cross-cache deadlock in ObservableCache

Fixes the cross-cache deadlock described in #1073 by using a queue to avoid holding a lock while firing downstream events.

## Problem

ObservableCache held its lock (`_locker`) while calling `_changes.OnNext()` to deliver notifications to subscribers. When a subscriber's callback propagated through a pipeline (Transform, PopulateInto, MergeMany) into a second cache, the callback needed that second cache's lock. If another thread was simultaneously writing to the second cache and its subscriber callback chained back to the first cache, both threads blocked permanently waiting for each other's lock.

This was confirmed through analysis of a production hang where 6 threads were deadlocked across multiple ObservableCache instances. The stacks showed the classic pattern: Thread A held CacheA._locker and was blocked on CacheB._locker, while Thread B held CacheB._locker and was blocked on CacheA._locker.

## Solution: Queue-Based Drain Pattern

The fix separates mutation from notification delivery. Instead of calling `OnNext` while holding the lock, writes now enqueue a notification item under the lock, then deliver it after releasing the lock. No lock is ever held when subscriber code runs, which eliminates the cross-cache deadlock entirely.

The core of this approach is a new `DeliveryQueue<TItem>` class that encapsulates the pattern.

### DeliveryQueue

`DeliveryQueue<TItem>` is a generic, reusable concurrency primitive in `src/DynamicData/Internal/DeliveryQueue.cs`. It manages a FIFO queue of items and a delivery token that ensures only one thread delivers at a time.

The public API is designed to make incorrect usage difficult:

- `AcquireLock()` returns a `ref struct ScopedAccess` that holds the caller's lock. Queue operations (`Enqueue`) go through this scoped handle. When it is disposed, the lock is released and pending items are delivered outside the lock. Because it is a `ref struct`, it cannot escape the calling method.
- `AcquireReadLock()` returns a `ref struct ReadOnlyScopedAccess` for inspecting queue state without mutation. It exposes a single property, `HasPending`, which reports whether items are queued or mid-delivery. Disposing releases the lock without triggering delivery. This is used by `Connect` and `Watch` to determine whether version filtering is needed.
- The delivery callback is a `Func<TItem, bool>` that returns `true` to continue or `false` to signal termination. When the callback returns `false`, the queue sets itself as terminated, clears remaining items, and stops. This gives the queue full ownership of its lifecycle.
- `IsTerminated` is a volatile flag safe to read from any thread.
- The delivery loop acquires the lock only for the dequeue step, then releases it before calling the delivery callback. If the callback throws, a `catch/throw` block resets the delivery token so the queue recovers on the next write.
- NET9+ uses `System.Threading.Lock` with `Lock.Enter()`/`Lock.Exit()`. Older TFMs use `Monitor.Enter()`/`Monitor.Exit()`. Both are defined under a single `#if` block for clarity.

### How ObservableCache Uses It

Write paths (`UpdateFromSource`, `UpdateFromIntermediate`, source subscription `OnNext`):
```csharp
using var notifications = _notifications.AcquireLock();
var changes = _readerWriter.Write(updateAction, previewHandler, _changes.HasObservers);
if (changes is not null)
{
    notifications.Enqueue(NotificationItem.CreateChanges(changes, _readerWriter.Count));
}
// ScopedAccess.Dispose releases lock, then delivers if this thread holds the delivery token
```

Terminal events (source completion, source error, dispose) go through the same queue as sentinel items. The delivery callback returns `false` for these, and the queue self-terminates. This means all terminal cleanup (completing subjects, disposing suspension tracker) happens through a single code path in `DeliverNotification`.

`Connect` and `Watch` use `DeliveryQueue.AcquireReadLock()` to atomically read queue state, take a snapshot, and subscribe (all under a single lock acquisition). When `ReadOnlyScopedAccess.HasPending` indicates notifications are queued or mid-delivery, the snapshot may contain data whose notification hasn't fired yet. To prevent duplicates, each enqueued change notification carries a monotonically increasing version (`_currentVersion`, incremented under lock). Before calling `_changes.OnNext()`, the delivery callback stamps the item's version into `_currentDeliveryVersion` via `Volatile.Write`. The subscriber's stream is wrapped with `.SkipWhile(_ => Volatile.Read(ref _currentDeliveryVersion) <= snapshotVersion)`, which drops notifications already covered by the snapshot. Once the first post-snapshot notification arrives, Rx's `SkipWhile` nulls out its predicate (zero ongoing cost). When no notifications are pending, the subscriber gets the raw `_changes` Subject with no filter overhead.

### Suspension

Suspension state is checked at delivery time via local helper methods `EmitOrSuspendChanges` and `EmitOrSuspendCount`. These check the live `AreNotificationsSuspended`/`IsCountSuspended` state under the lock at the moment of delivery, avoiding TOCTOU issues that would arise from capturing the state at enqueue time. A fast-path check on `_suspensionTracker.IsValueCreated` (a monotone volatile flag on `Lazy<T>`) skips the lock entirely when suspension has never been used.

`SuspensionTracker` no longer holds delegates into the queue. `ResumeNotifications` returns the accumulated changes; the caller enqueues them through `ScopedAccess`. The resume signal (`_areNotificationsSuspended.OnNext(false)`) is emitted after releasing the lock so that deferred `Connect`/`Watch` subscribers are activated outside the lock scope.

### NotificationItem

`NotificationItem` uses a `NotificationKind` enum (`Changes`, `CountOnly`, `Completed`, `Error`) and all instances are created through factory methods (`CreateChanges`, `CreateCountOnly`, `CreateCompleted`, `CreateError`). The delivery callback uses a switch on the kind for dispatch.

### Connect/Watch Snapshot-Duplicate Fix

The drain-outside-lock design introduced a race where `Connect()` could take a snapshot containing data whose notification was still queued or mid-delivery. When that notification later fired via `_changes.OnNext()`, the new subscriber saw duplicate Adds.

The fix uses version-tagged delivery: each enqueued change notification carries a monotonically increasing version (`_currentVersion`, incremented under lock). Before calling `_changes.OnNext()`, the delivery callback stamps the item's version into `_currentDeliveryVersion` via `Volatile.Write`. `CreateConnectObservable` and `CreateWatchObservable` capture `snapshotVersion = _currentVersion` under lock and, only when `DeliveryQueue.ReadOnlyScopedAccess.HasPending` is true, apply `.SkipWhile(_ => Volatile.Read(ref _currentDeliveryVersion) <= snapshotVersion)` to the `_changes` stream. Once the first post-snapshot notification arrives, `SkipWhile` nulls out its predicate (zero ongoing cost).

`ReadOnlyScopedAccess` is a new `ref struct` on `DeliveryQueue` that acquires the gate for inspection only (exposes `HasPending` but no mutation methods), and its `Dispose` releases the lock without triggering delivery.

## Testing

All existing tests pass, including all multi-threaded stress tests. 21 new tests across 3 fixtures, plus 1 existing test updated for timing reliability.

### DeliveryQueueFixture (14 tests)

Covers the new `DeliveryQueue<T>` concurrency primitive.

| Test | Description |
|------|-------------|
| `EnqueueAndDeliverDeliversItem` | Basic enqueue → deliver round-trip |
| `DeliverDeliversItemsInFifoOrder` | FIFO ordering guarantee |
| `DeliverWithEmptyQueueIsNoOp` | No delivery callback when queue is empty |
| `OnlyOneDelivererAtATime` | Delivery token serialization under concurrent writers |
| `SecondWriterItemPickedUpByFirstDeliverer` | Late-enqueued items delivered by the active deliverer |
| `ReentrantEnqueueDoesNotRecurse` | Enqueue from within delivery callback does not stack-overflow |
| `ExceptionInDeliveryResetsDeliveryToken` | Delivery token recovery after callback exception |
| `RemainingItemsDeliveredAfterExceptionRecovery` | Queue resumes after exception on next write |
| `TerminalCallbackStopsDelivery` | `return false` terminates queue and clears remaining items |
| `EnqueueAfterTerminationIsIgnored` | Writes after terminal callback are silently dropped |
| `IsTerminatedIsFalseInitially` | Initial state verification |
| `ConcurrentEnqueueAllItemsDelivered` | Stress: all items from concurrent writers are delivered |
| `ConcurrentEnqueueNoDuplicates` | Stress: no item delivered more than once |
| `ConcurrentEnqueuePreservesPerThreadOrdering` | Stress: per-writer FIFO ordering preserved under contention |

### SourceCacheFixture (3 tests)

Covers the cross-cache deadlock fix and snapshot-duplicate prevention.

| Test | Description |
|------|-------------|
| `DirectCrossWriteDoesNotDeadlock` 🐛 | **Proves original deadlock.** Bidirectional cross-cache writes via Filter → Transform → PopulateInto. 100 iterations, 2000 concurrent writes per iteration. Deadlocks on `main`. |
| `MultiCacheFanInDoesNotDeadlock` | Fan-in pattern stress test with subscribers that write back to other caches |
| `ConnectDuringDeliveryDoesNotDuplicate` 🐛 | **Proves snapshot-duplicate bug.** Slow subscriber blocks delivery of item1 while item2 is committed and queued. New Connect subscriber verifies each key appears exactly once. Validates version-tagged SkipWhile filtering. |

### SuspendNotificationsFixture (4 tests)

Covers the resume signal atomicity fix and suspend/resume ordering.

| Test | Description |
|------|-------------|
| `ResumeThenReSuspendDeliversFirstBatchOnly` | Deterministic: resume before re-suspend. Verifies exact message shape (2 messages with exact key ranges per message). |
| `ReSuspendThenResumeDeliversAllInSingleBatch` | Deterministic: re-suspend before resume. Verifies single changeset containing both batches. |
| `ConcurrentSuspendDuringResumeDoesNotCorrupt` | Stress: races resume vs re-suspend, 200 iterations. Verifies final state invariants. |
| `ResumeSignalUnderLockPreventsStaleSnapshotFromReSuspend` 🐛 | **Proves resume signal race.** Slow subscriber blocks delivery, creating a window for re-suspend + write. Deferred subscriber's snapshot must not contain data from the re-suspension. Fails without the fix (200 adds instead of 100). |

### MergeManyChangeSetsCacheSourceCompareFixture (1 modified test)

Existing test updated for queue-based delivery timing.

| Test | Change |
|------|--------|
| `MergeManyStress` | Added `Publish`/`Connect` and `LastOrDefaultAsync` await to ensure all notifications are delivered before assertions |

## Also Included

- **SwappableLock**: Added NET9+ `Lock` type overloads. Fixed `SwapTo(object)` to handle the case where the lock was initialized via `CreateAndEnter(Lock)`.
- **ExpireAfter**: Fixed a race condition where an item could be removed or updated by another thread between when its expiration was scheduled and when it fired. Also fixed `=` to `|=` for `haveExpirationsChanged` in the Update case to prevent silently dropping a prior change flag.

### CODE REVIEW

**Review by JakenVeina — CHANGES_REQUESTED** [review]
There's a critical flaw here, that I discussed in the corresponding issue: there's a race condition that allows for corruption of the change stream. If 2 edits are occurring in parallel, it's possible for edit A to release `_locker`, and then get suspended before it acquires `_notificationLocker`. Within this window, edit B can come in, make it's changes, and then acquire `_notificationLocker` before edit A.

There's a mechanism in the codebase already for coordinating lock handoff like this, where you need to acquire a second lock before releasing the first. I don't recall exactlybwhere it is or what it's called, but I believe I wrote it for one of the `.Filter()` operators. Lemme know if you have trouble finding it. I'm out of town, so it's tough for me to look just on my phone.

**Comment by dwcullop** [comment]
> There's a mechanism in the codebase already for coordinating lock handoff like this, where you need to acquire a second lock before releasing the first. 

Perfect.  Exactly what I needed to close the gap that two locks introduces.

**Comment by dwcullop** [comment]
I had to completely change tactics to get all the corner cases closed but this is working and the unit tests pass.  I updated the bug with a real repro, but it is kind of contrived.  It's just not easy to repro what I am getting to production without contriving it.

File: src/DynamicData/Cache/ObservableCache.cs [hunk 1]
```diff
@@ -246,8 +235,12 @@ private IObservable<IChangeSet<TObject, TKey>> CreateConnectObservable(Func<TObj
             {
                 lock (_locker)
                 {
+                    // Skip pending notifications to avoid duplicating items already in the snapshot.
+                    var skipCount = _notifications.PendingCount;
+
                     var initial = InternalEx.Return(() => (IChangeSet<TObject, TKey>)GetInitialUpdates(predicate));
-                    var changes = initial.Concat(_changes);
+                    var changesStream = skipCount > 0 ? _changes.Skip(skipCount) : _changes;
+                    var changes = initial.Concat(changesStream);
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData/Cache/ObservableCache.cs]: PendingCount removed.

File: src/DynamicData/Cache/ObservableCache.cs [hunk 2]
```diff
@@ -268,13 +261,16 @@ private IObservable<Change<TObject, TKey>> CreateWatchObservable(TKey key) =>
...
+
                     var initial = _readerWriter.Lookup(key);
                     if (initial.HasValue)
                     {
                         observer.OnNext(new Change<TObject, TKey>(ChangeReason.Add, key, initial.Value));
                     }
 
-                    return _changes.Finally(observer.OnCompleted).Subscribe(
+                    var changesStream = skipCount > 0 ? _changes.Skip(skipCount) : _changes;
+                    return changesStream.Finally(observer.OnCompleted).Subscribe(
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData/Cache/ObservableCache.cs]: PendingCount removed.

File: src/DynamicData/Cache/ObservableCache.cs [hunk 3]
```diff
@@ -289,77 +285,190 @@ private IObservable<Change<TObject, TKey>> CreateWatchObservable(TKey key) =>
...
+                NotificationItem.CreateChanges(resumedChanges, _readerWriter.Count, isSuspended: false, isCountSuspended: false),
+                countAsPending: true);
+        }
+
+        if (emitResume)
         {
-            Debug.Assert(_suspensionTracker.IsValueCreated, "Should not be Resuming Notifications without Suspend Count instance");
-            _suspensionTracker.Value.ResumeNotifications();
+            _suspensionTracker.Value.EmitResumeNotification();
         }
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData/Cache/ObservableCache.cs:446]: Fixed.

File: src/DynamicData.Tests/Cache/MergeManyChangeSetsCacheSourceCompareFixture.cs [hunk 4]
```diff
@@ -118,6 +118,11 @@ IObservable<MarketPrice> AddRemovePrices(Market market, int priceCount, int para
         }
         while (adding);
 
+        // Allow any in-flight notification deliveries to complete before checking results.
+        // With the queue-based drain pattern, Edit() returns after enqueueing but delivery
+        // may still be in progress on another thread. Give the drain a moment to finish.
+        await Task.Delay(250).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData.Tests/Cache/MergeManyChangeSetsCacheSourceCompareFixture.cs]: Fixed.

File: src/DynamicData/Cache/ObservableCache.cs [hunk 5]
```diff
@@ -188,178 +172,300 @@ internal void UpdateFromIntermediate(Action<ICacheUpdater<TObject, TKey>> update
...
+        if (!emitResume)
         {
-            Debug.Assert(_suspensionTracker.IsValueCreated, "Should not be Resuming Notifications without Suspend Count instance");
-            _suspensionTracker.Value.ResumeNotifications();
+            return;
         }
+
+        // Emit the resume signal under the lock to eliminate the race where a concurrent
+        // SuspendNotifications could produce overlapping emissions.
+        _suspensionTracker.Value.EmitResumeNotification();
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData/Cache/ObservableCache.cs]: This has been fixed.

File: src/DynamicData.Tests/Cache/SourceCacheFixture.cs [hunk 6]
```diff
@@ -188,4 +191,168 @@ public void StaticFilterRemove()
...
+
+        // Unblock the slow subscriber — delivery resumes, item2 delivered
+        connectDone.Set();
+        writeTask.Wait(TimeSpan.FromSeconds(5));
+        writeTask2.Wait(TimeSpan.FromSeconds(5));
+
+        // Each key should appear exactly once in the new subscriber's view
+        addCounts.GetValueOrDefault("k1").Should().Be(1, "k1 should appear once (snapshot only)");
+        addCounts.GetValueOrDefault("k2").Should().Be(1, "k2 should appear once, not duplicated from snapshot + queued delivery");
+    }
```
[the remark this answers is not in the document]
Comment (dwcullop) [reply src/DynamicData.Tests/Cache/SourceCacheFixture.cs:355]: Fixed.

**Comment by mycroes-nl** [comment]
I was expecting this to use System.Threading.Channels for a modern queuing implementation, but then again I'm not familiar with the internals of DynamicData at all so it might not fit in here.

**Comment by dwcullop** [comment]
> System.Threading.Channels

Interesting.  I'll see what that might get us.

**Comment by dwcullop** [comment]
> > System.Threading.Channels
> 
> Interesting. I'll see what that might get us.

They don't really fit our pattern.  It would mean making a lot of things async, which means more allocations on the hot path.  Instead, we'll use a custom, carefully crafted low-level class just for our purposes.  Harder to pull off, but less overhead and better performance for end users.

**Comment by dwcullop** [comment]
Closing this PR in favor of #1079.  It doesn't make sense to only do a half-fix.
